collectors/mmrepquota.go

```diff
@@ -59,102 +63,204 @@ type QuotaMetric struct {
+
+		GroupBlockUsage: prometheus.NewDesc(prometheus.BuildFQName(namespace, "group", "used_bytes"),
+			"GPFS group quota used", group_labels, nil),
+		GroupBlockQuota: prometheus.NewDesc(prometheus.BuildFQName(namespace, "group", "quota_bytes"),
+			"GPFS group block quota", group_labels, nil),
+		GroupBlockLimit: prometheus.NewDesc(prometheus.BuildFQName(namespace, "group", "limit_bytes"),
+			"GPFS group quota block limit", group_labels, nil),
+		GroupBlockInDoubt: prometheus.NewDesc(prometheus.BuildFQName(namespace, "group", "in_doubt_bytes"),
+			"GPFS group quota block in doubt", group_labels, nil),
+		GroupFilesUsage: prometheus.NewDesc(prometheus.BuildFQName(namespace, "group", "used_files"),
+			"GPFS group quota files used", group_labels, nil),
+		GroupFilesQuota: prometheus.NewDesc(prometheus.BuildFQName(namespace, "group", "quota_files"),
+			"GPFS group files quota", group_labels, nil),
+		GroupFilesLimit: prometheus.NewDesc(prometheus.BuildFQName(namespace, "group", "limit_files"),
+			"GPFS group quota files limit", group_labels, nil),
+		GroupFilesInDoubt: prometheus.NewDesc(prometheus.BuildFQName(namespace, "group", "in_doubt_files"),
+			"GPFS group quota files in doubt", group_labels, nil),
+
 		logger: logger,
 	}
 }
 
 func (c *MmrepquotaCollector) Describe(ch chan<- *prometheus.Desc) {
-	ch <- c.BlockUsage
-	ch <- c.BlockQuota
-	ch <- c.BlockLimit
-	ch <- c.BlockInDoubt
-	ch <- c.FilesUsage
-	ch <- c.FilesQuota
-	ch <- c.FilesLimit
-	ch <- c.FilesInDoubt
+	ch <- c.FilesetBlockUsage
+	ch <- c.FilesetBlockQuota
+	ch <- c.FilesetBlockLimit
+	ch <- c.FilesetBlockInDoubt
+	ch <- c.FilesetFilesUsage
+	ch <- c.FilesetFilesQuota
+	ch <- c.FilesetFilesLimit
+	ch <- c.FilesetFilesInDoubt
+
+	ch <- c.UserBlockUsage
+	ch <- c.UserBlockQuota
+	ch <- c.UserBlockLimit
+	ch <- c.UserBlockInDoubt
+	ch <- c.UserFilesUsage
+	ch <- c.UserFilesQuota
+	ch <- c.UserFilesLimit
+	ch <- c.UserFilesInDoubt
+
+	ch <- c.GroupBlockUsage
+	ch <- c.GroupBlockQuota
+	ch <- c.GroupBlockLimit
+	ch <- c.GroupBlockInDoubt
+	ch <- c.GroupFilesUsage
+	ch <- c.GroupFilesQuota
+	ch <- c.GroupFilesLimit
+	ch <- c.GroupFilesInDoubt
 }
 
 func (c *MmrepquotaCollector) Collect(ch chan<- prometheus.Metric) {
 	level.Debug(c.logger).Log("msg", "Collecting mmrepquota metrics")
 	collectTime := time.Now()
 	timeout := 0
 	errorMetric := 0
-	metrics, err := c.collect()
-	if err == context.DeadlineExceeded {
-		timeout = 1
-		level.Error(c.logger).Log("msg", "Timeout executing mmrepquota")
-	} else if err != nil {
-		level.Error(c.logger).Log("msg", err)
-		errorMetric = 1
+	metrics := []QuotaMetric{}
+
+	for _, quotaType := range *configMmrepquotaTypes {
+		metric, err := c.collect(fmt.Sprintf("-%c", quotaType))
+		metrics = append(metrics, metric...)
+
+		if err == context.DeadlineExceeded {
+			timeout = 1
+			level.Error(c.logger).Log("msg", "Timeout executing mmrepquota")
+		} else if err != nil {
+			level.Error(c.logger).Log("msg", err)
+			errorMetric = 1
+		}
 	}
```

Comment:
This should utilize go routines to run the collections in parallel.  Let me know if need help with that.  Something like this is initial idea:

```
quotaTypes := strings.Split(*configMmrepquotaTypes, ",")
for _, quotaType := range quotaTypes {
  wg.Add(1)
  quotaArg := quotaTypesMap[quotaType]
  go func(quotaArg string) {
    defer wg.Done()
    metric, err := c.collect(fmt.Sprintf("-%c", quotaArg))
    ... logic to append , catch errors , define metrics etc ...
  }
}
wg.Wait()
```

If you do the append you will have to use locks.  It might be simpler to define the Prometheus metrics inside the Go routine but either way works.

Another solution that might be cleaner is move the go routine and look into `collect()` and you will have to use locking to avoid race conditions when appending the metrics